pkg/controller/volume/vacprotection/vac_protection_controller.go

```diff
@@ -0,0 +1,403 @@
+
+	pvs, err := c.getPVsAssignedToVAC(vac.Name)
+	if err != nil {
+		logger.Error(err, "Error getting PVs assigned to VAC", "VAC", klog.KObj(vac))
+		return true
+	}
+	if len(pvs) > 0 {
+		return true
+	}
+
+	pvcs, err := c.getPVCsAssignedToVAC(vac.Name)
+	if err != nil {
+		logger.Error(err, "Error getting PVCs assigned to VAC", "VAC", klog.KObj(vac))
+		return true
+	}
+	if len(pvcs) > 0 {
+		return true
+	}
+	return false
+}
+
+// pvAddedUpdated reacts to vac added/updated events
+func (c *Controller) vacAddedUpdated(logger klog.Logger, obj interface{}) {
+	vac, ok := obj.(*storagev1alpha1.VolumeAttributesClass)
+	if !ok {
+		utilruntime.HandleError(fmt.Errorf("VAC informer returned non-VAC object: %#v", obj))
+		return
+	}
+	logger.V(4).Info("Got event on VAC", "VAC", klog.KObj(vac))
+
+	if protectionutil.NeedToAddFinalizer(vac, volumeutil.VACProtectionFinalizer) || protectionutil.IsDeletionCandidate(vac, volumeutil.VACProtectionFinalizer) {
+		c.queue.Add(vac.Name)
+	}
+}
+
+// pvcDeleted reacts to pvc deleted events
+func (c *Controller) pvcDeleted(logger klog.Logger, obj interface{}) {
+	pvc, ok := obj.(*v1.PersistentVolumeClaim)
+	if !ok {
+		utilruntime.HandleError(fmt.Errorf("PVC informer returned non-PVC object: %#v", obj))
+		return
+	}
+	logger.V(4).Info("Got event on PVC", "PVC", klog.KObj(pvc))
+	vacNames := getPVCReferencedVACNames(pvc)
+	for _, vacName := range vacNames {
+		c.queue.Add(vacName)
+	}
+}
+
+// pvcUpdated reacts to pvc updated events
+func (c *Controller) pvcUpdated(logger klog.Logger, old, new interface{}) {
+	oldPVC, ok := old.(*v1.PersistentVolumeClaim)
+	if !ok {
+		utilruntime.HandleError(fmt.Errorf("PVC informer returned non-PVC object: %#v", old))
+		return
+	}
+	newPVC, ok := new.(*v1.PersistentVolumeClaim)
+	if !ok {
+		utilruntime.HandleError(fmt.Errorf("PVC informer returned non-PVC object: %#v", new))
+		return
+	}
+
+	logger.V(4).Info("Got event on PVC", "PVC", klog.KObj(newPVC))
+
+	vavNames := sets.New(getPVCReferencedVACNames(oldPVC)...).Delete(getPVCReferencedVACNames(newPVC)...).UnsortedList()
+	for _, vacName := range vavNames {
+		c.queue.Add(vacName)
+	}
+}
+
+// pvUpdated reacts to pv updated events
+func (c *Controller) pvUpdated(logger klog.Logger, old, new interface{}) {
+	oldPV, ok := old.(*v1.PersistentVolume)
+	if !ok {
+		utilruntime.HandleError(fmt.Errorf("PV informer returned non-PV object: %#v", old))
+		return
+	}
+	newPV, ok := new.(*v1.PersistentVolume)
+	if !ok {
+		utilruntime.HandleError(fmt.Errorf("PV informer returned non-PV object: %#v", new))
+		return
+	}
+
+	logger.V(4).Info("Got event on PV", "PV", klog.KObj(newPV))
+	vavNames := sets.New(getPVReferencedVACNames(oldPV)...).Delete(getPVReferencedVACNames(newPV)...).UnsortedList()
+	for _, vacName := range vavNames {
+		c.queue.Add(vacName)
+	}
+}
+
+// pvDeleted reacts to pv deleted events
+func (c *Controller) pvDeleted(logger klog.Logger, obj interface{}) {
+	pv, ok := obj.(*v1.PersistentVolume)
```

Comment:
Should we delete the Indexers here for the vacNameKeyIndex?

Reply:
why to delete the indexer?

If an object is removed from the informer, we will never re-get the item via an indexer.

Reply:
And the entry of the indexer will be removed automatically? Or it will dangling around in some Get calls?

Reply:
> And the entry of the indexer will be removed automatically?

Yes. 

> Or it will dangling around in some Get calls?

No.